oximeter/db/src/oxql/plan/predicates.rs

```diff
@@ -0,0 +1,649 @@
+            pushed.push(disjunct.remove_datum(schema)?);
+            not_pushed.push(disjunct.only_datum(schema)?);
+        }
+
+        // We need to "compress" either side if it's an array of all Nones. That
+        // means none of disjuncts could be reordered, and so we'll need to
+        // elide the plan node entirely.
+        let pushed = if pushed.iter().all(Option::is_none) {
+            None
+        } else {
+            Some(Predicates::from(OptionalDisjunctions(pushed)))
+        };
+        let not_pushed = if not_pushed.iter().all(Option::is_none) {
+            None
+        } else {
+            Some(Predicates::from(OptionalDisjunctions(not_pushed)))
+        };
+        Ok(SplitPredicates { pushed, not_pushed })
+    }
+
+    pub(crate) fn to_required(&self) -> anyhow::Result<Vec<filter::Filter>> {
+        match self {
+            Predicates::Single(single) => {
+                single.simplify_to_dnf().map(|f| f.flatten_disjunctions())
+            }
+            Predicates::Disjunctions(disjuncts) => {
+                let mut out = Vec::with_capacity(disjuncts.len());
+                for disjunct in disjuncts.iter() {
+                    out.extend(
+                        disjunct
+                            .as_ref()
+                            .expect("Must be Some(_) here")
+                            .simplify_to_dnf()?
+                            .flatten_disjunctions(),
+                    );
+                }
+                Ok(out)
+            }
+        }
+    }
+
+    /// And this predicate with another.
+    ///
+    /// This will fail if there are different numbers of predicates in each
+    /// argument, since we won't know how to match up the predicates in that
+    /// case.
+    pub(crate) fn and(&self, other: &Self) -> anyhow::Result<Self> {
+        match (self, other) {
+            (Predicates::Single(left), Predicates::Single(right)) => {
+                Ok(Self::Single(left.and(right)))
+            }
+            (
+                Predicates::Single(single),
+                Predicates::Disjunctions(disjunctions),
+            )
+            | (
+                Predicates::Disjunctions(disjunctions),
+                Predicates::Single(single),
+            ) => {
+                anyhow::ensure!(
+                    disjunctions.len() == 1,
+                    "Can only merge together a list of predicates with a \
+                    single predicate if the list has length 1"
+                );
+                if let Some(right) = &disjunctions[0] {
+                    Ok(Self::Single(single.and(right)))
+                } else {
+                    Ok(Self::Single(single.clone()))
+                }
+            }
+            (
+                Predicates::Disjunctions(left),
+                Predicates::Disjunctions(right),
+            ) => {
+                anyhow::ensure!(
+                    left.len() == 1 && right.len() == 1,
```

Comment:
This should be `left.len() == right.len()`, shouldn't it? If not, update the docstring to indicate this only works for things that are effectively single predicates.

Reply:
I'm going to keep it this way for now. It's more conservative, but I don't think we need it to be more general and it's a lot harder to verify in that case.